lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java

```diff
@@ -0,0 +1,536 @@
+        return false;
+      }
+
+      @Override
+      boolean advanceExactWithinBlock(IndexedDISI disi, int target) throws IOException {
+        final int targetInBlock = target & 0xFFFF;
+        final int targetWordIndex = targetInBlock >>> 6;
+
+        // If possible, skip ahead using the rank cache
+        rankSkip(disi, target);
+
+        for (int i = disi.wordIndex + 1; i <= targetWordIndex; ++i) {
+          disi.word = disi.slice.readLong();
+          disi.numberOfOnes += Long.bitCount(disi.word);
+        }
+        disi.wordIndex = targetWordIndex;
+
+        long leftBits = disi.word >>> target;
+        disi.index = disi.numberOfOnes - Long.bitCount(leftBits);
+        return (leftBits & 1L) != 0;
+      }
+
+
+    },
+    ALL {
+      @Override
+      boolean advanceWithinBlock(IndexedDISI disi, int target) throws IOException {
+        disi.doc = target;
+        disi.index = target - disi.gap;
+        return true;
+      }
+      @Override
+      boolean advanceExactWithinBlock(IndexedDISI disi, int target) throws IOException {
+        disi.index = target - disi.gap;
+        return true;
+      }
+    };
+
+    /** Advance to the first doc from the block that is equal to or greater than {@code target}.
+     *  Return true if there is such a doc and false otherwise. */
+    abstract boolean advanceWithinBlock(IndexedDISI disi, int target) throws IOException;
+
+    /** Advance the iterator exactly to the position corresponding to the given {@code target}
+     * and return whether this document exists. */
+    abstract boolean advanceExactWithinBlock(IndexedDISI disi, int target) throws IOException;
+  }
+
+  /**
+   * If the distance between the current position and the target is > 8 words, the rank cache will
+   * be used to guarantee a worst-case of 1 rank-lookup and 7 word-read-and-count-bits operations.
+   * Note: This does not guarantee a skip up to target, only up to nearest rank boundary. It is the
+   * responsibility of the caller to iterate further to reach target.
+   * @param disi standard DISI.
+   * @param target the wanted docID for which to calculate set-flag and index.
+   * @throws IOException if a DISI seek failed.
+   */
+  private static void rankSkip(IndexedDISI disi, int target) throws IOException {
+    final int targetInBlock = target & 0xFFFF;       // Lower 16 bits
+    final int targetWordIndex = targetInBlock >>> 6; // long: 2^6 = 64
+
+    // If the distance between the current position and the target is < 8 longs
+    // there is no sense in using rank
+    if (targetWordIndex - disi.wordIndex < RANK_BLOCK_LONGS) {
+      return;
+    }
+
+    // Resolve the rank as close to targetInBlock as possible (maximum distance is 8 longs)
+    // Note: rankOrigoOffset is tracked on block open, so it is absolute (e.g. don't add origo)
+    final int rankIndex = targetInBlock >> RANK_BLOCK_BITS; // 8 longs: 2^3 * 2^6 = 512
+    disi.slice.seek(disi.rankOrigoOffset + rankIndex*Short.BYTES);
```

Comment:
Seeking back and forth between ranks and the bitset could be slow with NIOFSDirectory for the same reason as the jump table. Maybe we should put the rank index at the beginning of blocks and load it up-front in an in-memory short[]?

Reply:
This is a hard one and I am not sure we can find a single solution that works best for all scenarios. And there are so many possibilities:

1. Allocate a slice for DENSE-rank: We already use many slices.
1. Make DENSE-rank and block-jump share the same slice. This thrashes the block-jump cache when a DENSE-rank lookup is needed, but not the other way as the need for skipping to a new block means that the DENSE block has been exhausted.
1. Don't have a special slice for DENSE-rank. A DENSE block is about 2 * 8192 bytes, so jumps within the first half would not mean thrashing.
1. Load all 128 shorts up front: Wasted if skips are < 512 docIDs or there is only a single (or possibly very few) lookups in the DENSE block.
1. Increase `RANK_BLOCK_SIZE`: A fair suggestion as the current 1 rank/8 longs is somewhat arbitrary. The concrete choice comes down to how large the average advance is and we have little knowledge there (at least I don't). It is still "only" a mitigation of the potentially-wasted-overhead problem.
1. Use heuristics: Load the 128 shorts on first advance within the DENSE block, if the advance is > `RANK_BLOCK_SIZE` and < X, where we choose X so that we guess there will only be a single advance (maybe 32768?). Makes the logic more complicated and branchy.
1. Interleave the ranking with the bits. This would make the math more complicated and probably require division and/or hurt single-docID-stepping. I'll think a bit about this and see if there is a way to do it cleanly.

Reply:
> Load all 128 shorts up front: Wasted if skips are < 512 docIDs or there is only a single (or possibly very few) lookups in the DENSE block.

This would be wasted indeed, but I'd expect reading in a byte[256] to be super cheap?

Reply:
A 8th option could be to drop the rank index entirely and decrease the size of blocks so that within-block skipping wouldn't be necessary anymore.

Reply:
Sorry, yes, I should have distinguished `short` read one-at-a-time from `byte[256]`. If that is indeed super cheap then I am leaning towards #4 (with byte[256]-read) or #6. With #4 preferred for simplicity.

Decreasing the size of blocks is a valid possibility and it is cheaper than before as they do not need to be iterated for long advances. I am not too positive about this idea as it changes the dynamic for `SPARSE`, where there will be wasted bits in the docIDs, and as it increases the block jump-table length. Also, what should the block-size be to get to a "reasonable" `DENSE`? 1024 bits? 2048? I am afraid that going so low would make the relative overhead from switching blocks noticeable higher.

So, should I try and implement #4? It's quite simple to do.

Reply:
+1 to try and implement 4

Reply:
As stated on the JIRA the pre-loading of rank is now implemented. It had the secondary effect of making the code cleaner. Should we need to change to `#6` heuristics, it can be done without too much trouble.